go-controller/pkg/clustermanager/egressip_controller.go

```diff
@@ -1723,3 +1711,146 @@ func (eIPC *egressIPClusterController) removePendingOpsAndGetResyncs(egressIPNam
 	}
 	return resyncs, nil
 }
+
+// jsonPatchOperation contains all the info needed to perform a JSON path operation to a k8 object
+type jsonPatchOperation struct {
+	Operation string      `json:"op"`
+	Path      string      `json:"path"`
+	Value     interface{} `json:"value,omitempty"`
+}
+
+// patchEgressIP performs a patch operation on an EgressIP.
+// There are two possible patches operations.
+// 1. Mandatory, replace operation of egress IP status field. This allows us to
+// update only the status field, without overwriting any other. This is
+// important because processing egress IPs can take a while (when running on a
+// public cloud and in the worst case), hence we don't want to perform a full
+// object update which risks resetting the EgressIP object's fields to the state
+// they had when we started processing the change.
+// 2. Optional, add operation to its metadata.annotations field.
+func (eIPC *egressIPClusterController) patchEgressIP(name string, patches ...jsonPatchOperation) error {
+	klog.Infof("Patching status on EgressIP %s: %v", name, patches)
+	op, err := json.Marshal(patches)
+	if err != nil {
+		return fmt.Errorf("error serializing patch operation: %+v, err: %v", patches, err)
+	}
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		return eIPC.kube.PatchEgressIP(name, op)
+	})
+}
+
+// generateEgressIPPatches conditionally generates a mark patch if the mark doesn't exist. If it fails to allocate a mark,
+// log an error instead of failing because we do not wish to block primary default network egress IP assignments due to potential
+// mark range exhaustion. Primary default network egress IP currently does not utilize marks to config EgressIP.
+// Generating the status patch is mandatory
+func (eIPC *egressIPClusterController) generateEgressIPPatches(name string, annotations map[string]string,
```

Comment:
can you expand on why the mark logic is coupled with status changes?
I am confused as to why this is being called from every status update part.... I'd expect those bits to not change...

Reply:
> can you expand on why the mark logic is coupled with status changes?

its to ensure when theres any status changes theres always going to be a mark, its a no-op usually. I was too lazy to do a lookup first to see if we need to perform the op.

Reply:
> I don't understand this; do you mean the cluster default network ? (i.e. the kubernetes managed network)

yes

> And what do we do if/when we the exhaust the mark range ? IMHO if the workload has required egressIPs, we need to put it in a crash-loop if we cannot fulfill the user's intents.

it was a design choice - CDN pods selected by an EIP will continue to function as it doesnt depend on pkt marks but UDN pods will not. I didnt want to block CDN pods.